### PR TITLE
[parallel] Let WorkerRunner know about paths, so PHPStan and Rector can handle static reflection over them

### LINKED ISSUE #6903: [Parallel]: Class was not found while trying to analyse it
# Bug Report

| Subject        | Details                                                         |
| :------------- | :---------------------------------------------------------------|
| Rector version | 0.12.10            |

## Minimal PHP Code Causing Issue

When enable parallel option, it can got error:

```bash
 [ERROR] Could not process "tests/system/Files/FileCollectionTest.php" file, due to:                                    
         "System error: "Class FileCollectionTest was not found while trying to analyse it - discovering symbols is     
         probably not configured properly."                                                                             
         Run Rector with "--debug" option and post the report here: https://github.com/rectorphp/rector/issues/new". On 
         line: 45     
```

Step to reproduce:

1. Clone repo

```bash
git clone git@github.com:codeigniter4/CodeIgniter4.git
```

2. Run composer install

```bash
composer install
```

3. Open `rector.php`, add option parall:

```php
$parameters->set(Option::PARALLEL, true);
```

4. Run against `tests` directory:

```bash
vendor/bin/rector process tests
```

## Expected Behaviour

It should show `[OK] Rector is done!` without diff and error:

```bash
vendor/bin/rector process tests 
 257/257 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

                                                                                                                        
 [OK] Rector is done!                                                                                    
```

### PR DESCRIPTION
**Pull request by TomasVotruba**
Replaces https://github.com/rectorphp/rector-src/pull/1613 (closes https://github.com/rectorphp/rector-src/pull/1613)

Fixes https://github.com/rectorphp/rector/issues/6903

### CODE REVIEW

File: e2e/no-parallel-reflection-resolver/rector.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+    $parameters->set(Option::PARALLEL, false);
+    $parameters->set(Option::PATHS, [
+        __DIR__ . '/src/',
+    ]);
+
+    $services = $containerConfigurator->services();
+    $services->set(RemoveUnusedPrivatePropertyRector::class);
+};
+
```
Comment (TomasVotruba) [inline e2e/no-parallel-reflection-resolver/rector.php:20]: I added one more test case, with parallel = false, so we know the results are for both identical.

File: e2e/e2eTestRunner.php [hunk 2]
```diff
@@ -27,19 +35,19 @@
...
-    echo "\nEXPECTED:\n";
-    var_dump($expectedOutput);
-    echo "\nACTUAL:\n";
-    var_dump($output);
-    exit(1);
-}
+// print color diff, to make easy find the differences
+$consoleDiffer = new ConsoleDiffer(new Differ(), new ColorConsoleDiffFormatter());
+$diff = $consoleDiffer->diff($output, $expectedOutput);
+$symfonyStyle->writeln($diff);
```
Comment (TomasVotruba) [inline e2e/e2eTestRunner.php:52]: The former report only showed full texts and reported "some difference somewhere".

Now the diff only shows which exact line is different :+1:  e.g.

image

File: packages/Parallel/WorkerRunner.php [hunk 3]
```diff
@@ -30,11 +32,15 @@ public function __construct(
...
         private readonly CurrentFileProvider $currentFileProvider,
         private readonly PhpFileProcessor $phpFileProcessor,
+        private readonly NodeScopeResolver $nodeScopeResolver,
+        private readonly DynamicSourceLocatorDecorator $dynamicSourceLocatorDecorator,
     ) {
     }
 
     public function run(Encoder $encoder, Decoder $decoder, Configuration $configuration): void
     {
+        $this->dynamicSourceLocatorDecorator->addPaths($configuration->getPaths());
```
Comment (TomasVotruba) [inline packages/Parallel/WorkerRunner.php:42]: The 1st missing part was paths here. That's how we can tell Rector and PHPStan to use static reflection on specific directories

File: packages/Parallel/WorkerRunner.php [hunk 4]
```diff
@@ -67,6 +73,9 @@ public function run(Encoder $encoder, Decoder $decoder, Configuration $configura
             $errorAndFileDiffs = [];
             $systemErrors = [];
 
+            // 1. allow PHPStan to work with static reflection on provided files
+            $this->nodeScopeResolver->setAnalysedFiles($filePaths);
+
```
Comment (TomasVotruba) [inline packages/Parallel/WorkerRunner.php:78]: The 2nd missing part was here. Now scope analyzer can work with provided paths.

File: e2e/parallel-reflection-resolver/src/NamespacedSomeClassFound.php [hunk 5]
```diff
@@ -0,0 +1,10 @@
+<?php
+
+declare(strict_types=1);
+
+namespace E2e\Parallel\Reflection\Resolver;
```
Comment (samsonasik) [inline e2e/parallel-reflection-resolver/src/NamespacedSomeClassFound.php:5]: The namespace need to be in composer.json dev-autoload to verify that namespaced class exists.

Comment (TomasVotruba) [reply e2e/parallel-reflection-resolver/src/NamespacedSomeClassFound.php:5]: Ok, I'll add it :+1:

**Comment by TomasVotruba** [comment]
@samsonasik Ready for review and merge :+1:

**Comment by samsonasik** [comment]
@TomasVotruba let's give it a try and let's see if we can add more directory to list of downgrade parallel.

**Review by samsonasik — APPROVED** [review]

**Pull request merged**